dfu/boot: double swap on interrupted revert bugfix adaptation

nvlsianpu opened the pull request:
Mcuboot changed TLV in PR: Fix double swap on interrupted revert
(https://github.com/JuulLabs-OSS/mcuboot/pull/485)

Above bugfix changes a little way for upgrade request.

This path introduces re-formatted original mcuboot bootutil_misc.c
code as much as it was reasonable which includes the bugfix
adaptation and support for devices witch have bite erased sate different
than 1 as well.

merge after https://github.com/JuulLabs-OSS/mcuboot/pull/485

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
dfu/boot: double swap on interrupted revert bugfix adaptation

Mcuboot changed TLV in PR: Fix double swap on interrupted revert
(https://github.com/JuulLabs-OSS/mcuboot/pull/485)

Above bugfix changes a little way for upgrade request.

This path introduces re-formatted original mcuboot bootutil_misc.c
code as much as it was reasonable which includes the bugfix
adaptation and support for devices witch have bite erased sate different
than 1 as well.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
cc @utzig

utzig reviewed (approved):
LGTM, a few nitpicks...

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -158,14 +175,100 @@ static const struct boot_swap_table boot_swap_tables[] = {
+	return BOOT_FLAG_SET;
+}
+
+/* this function should be mowed to flash_area api in future */
```
I have issue for that already. Now have argument for it.

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -158,14 +175,100 @@ static const struct boot_swap_table boot_swap_tables[] = {
+	return ERASED_VAL;
+}
+
+/* this function should be mowed to flash_area api in future */
```
I have issue for that already. Now have argument for it.

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -267,6 +383,25 @@ static int boot_magic_write(u8_t bank_id)
+
 static int boot_read_v1_header(u8_t area_id,
 			       struct mcuboot_v1_raw_header *v1_raw)
 {
```
@mbolivar As I remember you have added this. So can you give us an idea what should be done with this? - for sure any changes to this should go in different PR.

mbolivar replied:
My idea at the time was that zephyr applications should be able to work with multiple versions of MCUboot. I.e. modern zephyr, old MCUboot, should continue to work forever, since MCUboot is considered non-upgradable.

Think of a device that needs to live for years.

nvlsianpu commented:
@mbolivar Pleas review.

mbolivar reviewed (approved):
Was waiting for @utzig to review and for you to respond :)

If he's happy, I'm happy. I don't understand the issue well enough to say but the changes seem harmless.

nvlsianpu commented:
@d3zd3z Any chance that you can look at this soon?

carlescufi reviewed (changes requested):
Blocking this until it is made configurable

nvlsianpu committed:
dfu/boot: make swap_type optional

Swap_type field implementation is the new mcuboot's trailer field.
This patch makes using of it optional which keeps possibility of being
compatible with older versions of MCUBoot.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
